### DISCUSSION TITLE
My personal CSS customizations!

### DISCUSSION THREAD

**Discussion by djsime1** [Zettlr/Zettlr#2863] [Show and tell]
👋 Hey folks, I recently switched from Obsidian to Zettlr and while it's fine out of the box, I felt it could have used a few UI improvements. Thus, I authored some CSS customizations. Nothing extreme in my opinion.  
- More info on enabling custom CSS: [Zettlr docs](https://docs.zettlr.com/en/core/custom-css/)
- The source code: [GitHub Gist](https://gist.github.com/djsime1/2cffa7bada88573d12523ed0031eb27c)
- List of changes can be found in the Gist.

💻 In theory, *this should work regardless of your operating system or settings.*
🎉 **Suggestions are welcome in the comments here or on the Gist!**
✏ I'll update the Gist with any new changes (and hopefully remember to comment here too)
🖼 Preview screenshot below. 👇

![image](https://user-images.githubusercontent.com/8518150/143661724-f155a589-5c6f-4a53-ad8f-441da107c79c.png)
# [**_👉 Download here! 👈_**](https://gist.github.com/djsime1/2cffa7bada88573d12523ed0031eb27c)

**Comment by dajare** [comment]
It would be great to see a collection of these develop. :) Or is there already a listing somewhere? (Also left a comment on the gist.)

**Reply by djsime1** [reply]
To my knowledge, there isn't a collection _yet._

**Reply by nathanlesage** [reply]
There's not a list of these, but … idk, we might just start a repo and add an ability to pull them in for users …? Definitely not a priority, but we can certainly discuss that! :)

**Reply by dajare** [reply]
> There's not a list of these, but … idk, we might just start a repo and add an ability to pull them in for users …? Definitely not a priority, but we can certainly discuss that! :)

Even just having this thread in the Discussions is a help. :+1: (So certainly "not a priority", agreed!)

**Reply by newhinton** [reply]
I disagree with the git-repo for themes for multiple reasons:

1. Git is complicated. What i mean is that a normal user that just wants a nice app probably does not now how to get a theme out of a repo.
2. Single point of failure. If for whatever reason that repo goes, it will be hard to find themes again. Making it distributed solves that issue. Also a user is not reliant on someone else for their own themes. 
3. Maintenance: A repo needs to be maintained. If the theme-format changes, the maintainer would need to update ALL themes, or have a completely messy theme-repository, which is not desirable.
4. Distribution: Assuming a lot of people add themes, this can get "big" fast. We dont need 99% of those files on every users system, so pulling a theme should be manual anyway and then we are back at point 1. (I mean its just a bit of text, but still)


To clarify: i am not against a central place for themes, there could be a marketplace of some sort, but it should not rely on a singular repository to store theme-data. 

As outlined in #2861 i think a container-format would be appropriate, which then can be linked to some sort of central repository. The theme itself should be standalone somewhere else, like a github-release or a link to somewhere arbitrary. This way, even if the central repository is broken, all themes are still where they should be. It is also easier to distribute a theme without having to submit it to the central repository.

I highly appreciate any critique and comments on my idea!

**Reply by nathanlesage** [reply]
I did not intend to imply that users would have to actually work with the repo. I had something in mind as outlined below:

1. Integrate some form of theme "marketplace" in the app. This would then just download a metadata file (can be generated automatically) to present to the users.
2. Users could then just click a button which would initiate a download onto the computer (just that file, not the repo) and keep it updated.
3. Users could also dump theme files into the same folder, separately (think: Wordpress themes)
4. If the repo goes down, the themes are still there, just not updated, and by keeping manual ways of doing that, this should pose no problem

However, I agree in that it's some effort to maintain the repo itself (similar to plugins). However, I don't yet see an efficient way of having all of this distributed decentral in a way that's still easy for end users. I will have a look at the issue now and see if there's more in there, but feel free to comment here!

**Reply by newhinton** [reply]
Oh! Then i misunderstood you, i thought you wanted to put actual themes (css,metadata) in a repository. What you suggest is what i would like to see aswell, and more or less how many other projects do it aswell

**Reply by nathanlesage** [reply]
> i thought you wanted to put actual themes (css,metadata) in a repository

This *is* what I wanted to do, but I didn't want to download that full repo, because that would be weird :D

**Comment by Redsandro** [comment]
Very interesting. I like (and borrowed) that gradient css svg trick.

**Comment by IgnacioHeredia** [comment]
In the spirit of using this thread as a temporary repository for customizations, I leave a link to my [CSS file](https://gist.github.com/IgnacioHeredia/df82281250eacee16fd5eb644fbeba4e).

My customizations focus mainly on the editor:
![image](https://user-images.githubusercontent.com/16413205/186988225-eba67024-37ab-4250-8dfa-59623d9e87ff.png)

**Reply by Redsandro** [reply]
Those tags! Awesome. Thanks for the tip @IgnacioHeredia, I now have pretty tags too. :star_struck: 

You can improve the CSS for frontmatter users by adding `.zkn-tag` like so:

```css
/* Generic tags */
body.dark #editor .CodeMirror .cm-zkn-tag,
body.dark #editor .CodeMirror .zkn-tag {
   background-color: rgba(275,171,171, 1);
   color: rgba(138,0,0, 1);
   padding: 2px;
   border-radius:5px;
}
```

![image](https://user-images.githubusercontent.com/1702193/169815033-0988d025-461c-4c3e-841c-9d0662c0a1a4.png)

If preferred, you can add the # for visual purposes like so:

```css
body.dark #editor .CodeMirror .zkn-tag:before {
    content: "#";
}
```

![image](https://user-images.githubusercontent.com/1702193/169815398-1d30caf0-7ded-4d31-8b7b-918941c79abc.png)

They cannot be color coded because the named classes are missing. Might be worth an issue.

**Reply by IgnacioHeredia** [reply]
> You can improve the CSS for frontmatter users by adding .zkn-tag like so:

I don't use frontmatters, but thanks for sharing! :) Refining the refinement, you can make the CSS more concise with  the `is` idiom:
```css
body.dark #editor .CodeMirror :is(.cm-zkn-tag, .zkn-tag) {
   background-color: rgba(275,171,171, 1);
   color: rgba(138,0,0, 1);
   padding: 2px;
   border-radius:5px;
}
```

> They cannot be color coded because the named classes are missing. Might be worth an issue.

A few hours ago, I opened an [issue](https://github.com/Zettlr/Zettlr/issues/3495) about the missing first letter in the classes names. Maybe this can be mentionned there!

**Reply by nathanlesage** [reply]
Alongside the missing-letter bug I now also fixed this. Now you will have two class names for those tags: First `cm-tkn-tag-thetag` for any tags within the document, and then the same class, but without `cm-` for the same tags in the frontmatter. CodeMirror automatically prepends this `cm` but I figured this way if you want both tags to show the same, you just have to define two classes, and those who actually want to style both differently, can do so with ease.

**Reply by WolfgangDpunkt** [reply]
Has anyone here managed to adjust the spacing in bullet point lists?
It bothers me that between the individual list entries the spacing appears nearly as large as the normal line spacing. See screenshot:

<img width="803" alt="grafik" src="https://user-images.githubusercontent.com/35533844/181353608-ee3de762-7d97-462e-93ae-e0f01a719263.png">

However, I can't select the list items via CSS, since they are not real `<li></li>` elements. They don't seem to differ from normal text lines in the source code.

The only thing that seems to be selectable via CSS are the bullet point smybols: `.cm-formatting-list-ul`

Any advice would be appreciated, as I really like the clarity of bullet point lists in text editors.

**Reply by do-not-perceive-me** [reply]
you can increase the spacing with this code, but I can't figure out how to decrease the spacing, because that's set by the text line

```
.cm-variable-2{
    line-height: 110%
}
```

**Reply by sensologica** [reply]
@WolfgangDpunkt
I've been trying to achieve this, too, but bumped up against the same limitation. It looks like there is no selector for list items.

@do-not-perceive-me
Unfortunately, if the line wraps then the height between the first and the wrapped part of the line is also increased.

Ideally there would be a way to add vertical padding or margin to list items themselves, but that doesn't seem to work. My guess is that this is not allowed by CodeMirror?
